Write backwards compatible row group statistics (#3526)

tustvold opened the pull request:
# Which issue does this PR close?

<!--
We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.
-->

Closes #3526
Closes #799

# Rationale for this change
 
<!--
Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.
-->

This helps with compatibility with pyarrow, which doesn't understand min_value yet https://github.com/apache/arrow/issues/13976

# What changes are included in this PR?

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

# Are there any user-facing changes?


<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!---
If there are any breaking changes to public APIs, please add the `breaking change` label.
-->

tustvold commented:
Test failures fixed in #3528

alamb commented:
I think this may also fix https://github.com/apache/arrow-rs/issues/799 filed by @tfiasco and which @pjmore  also commented on

alamb commented on parquet/src/file/statistics.rs:
```diff
@@ -329,6 +331,12 @@ impl Statistics {
     }
 
+    /// Returns `true` if the statistics are backwards compatible with the
+    /// deprecated `min` and `max` fields
```
```suggestion
    /// deprecated `min` and `max` fields in parquet version `1.0` spec
```

alamb commented on parquet/src/column/writer/mod.rs:
```diff
@@ -817,13 +817,16 @@ impl<'a, E: ColumnValueEncoder> GenericColumnWriter<'a, E> {
                 self.column_metrics.num_column_nulls,
                 false,
             );
+            let statistics = statistics
```
```suggestion
            // Some common readers only support "old" statistics format
            // so write them out if possibe. See 
            // https://github.com/apache/arrow-rs/issues/799
            let statistics = statistics
```

alamb commented on parquet/src/column/writer/mod.rs:
```diff
@@ -1893,6 +1896,7 @@ mod tests {
     fn test_bool_statistics() {
         let stats = statistics_roundtrip::<BoolType>(&[true, false, false, true]);
         assert!(stats.has_min_max_set());
+        assert!(!stats.is_min_max_backwards_compatible());
```
is it just that bool is not backwards compatibel?

tustvold replied:
Yup, added a doc to that effect

alamb commented on parquet/src/file/statistics.rs:
```diff
@@ -252,10 +252,12 @@ pub fn to_thrift(stats: Option<&Statistics>) -> Option<TStatistics> {
-        thrift_stats.min = min;
-        thrift_stats.max = max;
-    } else {
+    if stats.is_min_max_backwards_compatible() {
```
```suggestion
    // Copy to min/max values from parquet 1.0 for compatibility
    // with other writers
    if stats.is_min_max_backwards_compatible() {
```

alamb commented on parquet/src/file/statistics.rs:
```diff
@@ -405,7 +413,13 @@ pub struct ValueStatistics<T> {
+    /// `min_value` and `max_value`
     is_min_max_deprecated: bool,
+
+    /// If `true` should always write deprecated `min` and `max` fields
```
is this in addition to `min_value` and `max_value`?

I found it hard to understand how these two flags were supposed to interact

alamb commented on parquet/src/file/statistics.rs:
```diff
@@ -423,6 +437,15 @@ impl<T: ParquetValueType> ValueStatistics<T> {
+        }
+    }
+
+    /// Set whether to write the deprecated `min` and `max` fields
```
```suggestion
    /// Set whether to write the deprecated `min` and `max` fields
    /// for compatibility with older parquet writers
```

alamb commented on parquet/src/file/statistics.rs:
```diff
@@ -478,6 +501,12 @@ impl<T: ParquetValueType> ValueStatistics<T> {
         self.is_min_max_deprecated
     }
+
+    /// Returns `true` if the statistics are backwards compatible
```
is it that the statistics are backwards compatible or they are desired to be?

alamb reviewed (approved):
LGTM -- thank you @tustvold . I think this will be a very nice feature addition for othr users. 

I had a bunch of naming / doc / style suggestions but nothing required I don't think

I also tested the reproducer from https://github.com/apache/arrow-rs/issues/799 from @tfiasco and the min/max values are properly written 👍 


```
In [3]: 
   ...: import pyarrow.parquet as pq
   ...: f = pq.ParquetFile('/tmp//test.parquet')
   ...: print(f.metadata.row_group(0).column(0).statistics)
<pyarrow._parquet.Statistics object at 0x110118c70>
  has_min_max: True
  min: 1
  max: 5
  null_count: 0
  distinct_count: 0
  num_values: 5
  physical_type: INT32
  logical_type: None
  converted_type (legacy): NONE

In [4]: 
```

tustvold commented:
I've updated the docs to hopefully be a bit clearer, PTAL.

I dropped the mentions of parquet v1, as I'm not entirely sure when the deprecation actually happened - the parquet changelog isn't particularly clear on this aspect...

alamb commented:
Thanks @tustvold
